### ISSUE TITLE
Auto-import consistently failing after a few minutes (happening in vscode, nightly ts)

### ISSUE THREAD

**Issue by jscheel** [microsoft/TypeScript#48880]
# Bug Report

I'm sorry, this may not be a very good bug report, but auto-import consistently stops working after a few minutes. I have posted a bit of logging below.

### 🔎 Search Terms

Auto-import crashes, Auto import failing, Auto import broken
### 🕗 Version & Regression Information

- None of these are applicable. I am using typescript nightly in vscode, and it has happened on an off for a long time.

### 🙁 Actual behavior

After a few minutes of work, auto importing stops working. If restart the tsserver in vscode the auto imports work again on the same line of code. This happens repeatedly after 10-15 minutes of coding. Even when importing the same thing in a different file. I noticed this in this logs:

```
Info 737  [14:16:50.020] getCompletionData: Get current token: 0.011509895324707031
Info 738  [14:16:50.020] getCompletionData: Is inside comment: 0.0037658214569091797
Info 739  [14:16:50.020] getCompletionData: Get previous token: 0.04294610023498535
Info 740  [14:16:50.020] getCompletionsAtPosition: isCompletionListBlocker: 0.01444697380065918
Info 741  [14:16:50.025] AutoImportProviderProject: attempted to add more than 10 dependencies. Aborting.
Info 742  [14:16:50.026] getExportInfoMap: cache hit
Info 743  [14:16:50.027] AutoImportProviderProject: attempted to add more than 10 dependencies. Aborting.
Info 744  [14:16:50.034] collectAutoImports: resolved 0 module specifiers, plus 0 ambient and 0 from cache
Info 745  [14:16:50.034] collectAutoImports: response is complete
Info 746  [14:16:50.034] collectAutoImports: 6.69482684135437
Info 747  [14:16:50.034] getCompletionData: Semantic work: 14.261340141296387
Info 748  [14:16:50.045] getCompletionsAtPosition: getCompletionEntriesFromSymbols: 11.098047971725464
```

As you can see, it does find results `attempted to add more than 10 dependencies`, but then it bails out with `no suggestions` in the vscode intellisense dropdown.

### 🙂 Expected behavior

Auto import works consistently.

**Comment by andrewbranch** [comment]
The bit you’ve keyed in on in the logs is a red herring. We need a repro and/or a full server log.

**Comment by flupke** [comment]
This happens to me constantly too, here is a log demonstrating the issue:

[tsserver-autoimport-failure-before-restart.log](https://github.com/microsoft/TypeScript/files/8590896/tsserver-autoimport-failure-before-restart.log)

1. At line 23006 I try to complete `OperationsCach` to trigger an auto-import and it fails
2. I restart the server
3. At line 25253 the auto import is successful

**Comment by jscheel** [comment]
@andrewbranch is @flupke's log enough to go off of, or do you need more from me?

**Comment by andrewbranch** [comment]
This isn’t a TSServer log; it’s VS Code’s client log from communicating with TS Server. It doesn’t contain any useful info on its own. See https://github.com/microsoft/TypeScript/wiki/Getting-logs-from-TS-Server-in-VS-Code for instructions on how to collect a log. @jscheel no log will ever be as good as a full repro.

**Comment by flupke** [comment]
Here are TSServer logs of an auto import failure:
[fail-autoimport.zip](https://github.com/microsoft/TypeScript/files/8613834/fail-autoimport.zip)
(look for `getTypedLayerDe\b` near the end)
And the same auto completion after a restart:
[success-autoimport.zip](https://github.com/microsoft/TypeScript/files/8613841/success-autoimport.zip)

**Comment by andrewbranch** [comment]
There’s no error in the logs, so unfortunately it doesn’t give me anything to go on. This isn’t going to be actionable without either (a) a way I can reproduce the issue on my own machine, or (b) hopping on a call and LiveShare debugging a repro on your machine.

**Comment by jscheel** [comment]
I'm not sure I can put together a reproducible case, since I don't have a clear pattern and inside a very large project. I can provide a bit more info: it just happened when I had two files open in vscode. Intellisense brought up the appropriate import on one file, which I then inserted and saved. Then I switch to the other open file, and it couldn't find anything. I'm definitely down for a liveshare, if that would help. It happens enough that we should be able to trigger it.

**Comment by andrewbranch** [comment]
And to clarify, once it can’t find the auto-import, you restarted TS Server without changing anything else and then it appeared? There are lots of reasons why an auto-import might not show up, but if you can fix it just by restarting the server it’s definitely a bug.

**Comment by flupke** [comment]
It seems to often happen to me after switching to a new file too. I'm also working on a fairly large project and I did not notice a clear pattern yet.

**Comment by andrewbranch** [comment]
> it has happened on an off for a long time

Roughly how long is a long time?

Was the imported file part of a different project in a project references setup?

> It seems to often happen to me after switching to a new file too.

By "new file" do you mean a file you just created or an existing file you just opened?

**Comment by flupke** [comment]
> And to clarify, once it can’t find the auto-import, you restarted TS Server without changing anything else and then it appeared? There are lots of reasons why an auto-import might not show up, but if you can fix it just by restarting the server it’s definitely a bug.

Yep, just restarting the server and triggering the exact same completion at the same spot is enough to fix the issue.

**Comment by jscheel** [comment]
> And to clarify, once it can’t find the auto-import, you restarted TS Server without changing anything else and then it appeared? There are lots of reasons why an auto-import might not show up, but if you can fix it just by restarting the server it’s definitely a bug.

Yep, just restarted ts server and boom, it popped up

**Comment by andrewbranch** [comment]
Next time that happens, instead of restarting the server, can you try going to a different file in the same tsconfig.json project (this won’t work if you go to a file in another composite project), trigger completions there (in a way that gets auto-imports, i.e. not a property access), and then go back to the original file and try again?

**Comment by flupke** [comment]
> 
> Was the imported file part of a different project in a project references setup?
> 
> 

It's in the same project (same tsconfig), part of a mono repo set up with yarn v3 workspaces.

> 
> > It seems to often happen to me after switching to a new file too.
> 
> 
> 
> By "new file" do you mean a file you just created or an existing file you just opened?

I mean opening an existing file in the same project (i.e. under the same workspace).

**Comment by andrewbranch** [comment]
Also when it happens, check the Output panel of the Extension Host in VS Code and make sure you don’t see any errors—I didn’t see any in @flupke’s logs but it’s possible the two of you are experiencing different root causes. If there’s an error stack it will obviously narrow down where to look a lot.

**Comment by flupke** [comment]
> Next time that happens, instead of restarting the server, can you try going to a different file in the same tsconfig.json project (this won’t work if you go to a file in another composite project), trigger completions there (in a way that gets auto-imports, i.e. not a property access), and then go back to the original file and try again?

I tried that and it fixes auto import without a restart :)

**Comment by flupke** [comment]
Almost all the code base uses named exports so it definitely happened on them. I *think* it also happened on a default export once or twice but I'm not sure. And we don't use `export =` exports or re-exports.

**Comment by andrewbranch** [comment]
So it was an export from your own TS code, not a node_module... and you’re sure the export isn’t part of a different referenced project (belongs to another tsconfig.json referenced in the importing file’s tsconfig.json `"references"`)?

**Comment by flupke** [comment]
Yes I'm sure all files are under the same `tsconfig.json`, we don't have `references` yet, just set up the `workspaces` in the top-level `package.json`.

And yes it only happens on our own imports. I can't tell for node_modules imports, I have disabled them because I'm in the process of typing the code base and they were adding too much noise at this point. The code base was converted from javascript with https://github.com/airbnb/ts-migrate (so it's full of `any` for the moment), if that matters.

**Comment by andrewbranch** [comment]
Once you observe a broken auto-import and restart the server, are you able to take specific steps to make it break again?

**Comment by flupke** [comment]
> Once you observe a broken auto-import and restart the server, are you able to take specific steps to make it break again?

I haven't found a clear pattern in the failures yet, but next time it happens I will try to close/reopen the file in which it happened, because it may be possible it happens only on files that were just opened.

**Comment by andrewbranch** [comment]
How often does this happen? Like multiple times a day? I might send you a custom TS build with additional logging.

**Comment by flupke** [comment]
It happens quite often, yesterday maybe 10-20 times (today much less but I think it's just because of my progress on the typing process).

Feel free to send me a custom build! I wish I could help more but I can't share the code base I'm working on.

**Comment by andrewbranch** [comment]
Ok I’m looking at useful places to add logs and I actually have a very strong hypothesis of what’s going wrong... I’m gonna skip the logging and have you try a most likely fixed build 😅

**Comment by jscheel** [comment]
I just had it happen on a file. I switch to another file, added the same method, and it found the import. Then I switched back to the original file, and it found the import there as well.

**Comment by andrewbranch** [comment]
@jscheel if my hypothesis and fix at #48979 is correct, you’ll see a response that looks something like `{"seq":0,"type":"response","command":"completionInfo","request_seq":290,"success":true,"performanceData":{"updateGraphDurationMs":48.1666110008955},"body":{"canceled":true}}` in your TS Server log shortly before things go off the rails.

**Comment by jscheel** [comment]
@andrewbranch I don't see it right now, because I had my logging paired down, but I seem to recall seeing that the other day.

**Comment by andrewbranch** [comment]
I’m 100% sure that cancellation here _can_ corrupt the cache and that it happened in @flupke’s failure log, so the case is all but closed.

**Comment by andrewbranch** [comment]
In ~11 hours you’ll be able to try typescript@4.7.0-dev.20220506 and I expect that should do better. Let me know if not. Thanks for answering a million questions!

**Comment by jscheel** [comment]
@andrewbranch awesome, super excited to try it out! Thank you so much for sticking with @flupke and I, even when we couldn't provide the best debugging details.

**Comment by flupke** [comment]
@andrewbranch I didn't use the feature much today, but it's been working flawlessly with typescript@4.7.0-dev.20220506 so far! :clap: :tada:

**Comment by jscheel** [comment]
Sure thing. Also, at least for me, I'm not doing anything wonky with
composite projects. Just opening the dir in vscode via `code .`

**Comment by cilvet** [comment]
Hi! I've been having trouble with autocomplete lately and I've stumbled upon this thread, which is the only one that pops up when looking for "AutoImportProviderProject: attempted to add more than" besides the code itself xD

The thing is, I was having trouble with autocomplete in a new project with lots of dependencies, and I tried going into the code and changing the hardcoded maxDependencies from 10 to 100, which solved the problem at the cost of extremely poor performance and constant process kills (which is to be expected).

I checked a different project in which autocomplete works well, and I see that it's loading 6 dependencies. I don't quite understand what might be happening here, I expect the ts server to work under any amount of dependencies.

I feel like there is something here I'm not understanding correclty
